### PR TITLE
Add OMS-Auditd-Plugin (auoms)

### PR DESCRIPTION
**Pull request by taglines**
See also corresponding pull-requests for Build-OMS-Agent-for-Linux and PowerShell-DSC-for-Linux as well as code committed to master branch of OMS-Auditd-Plugin (LCA approved, but not yet public)

### CODE REVIEW

**Comment by robbiezhang** [comment]
@Microsoft/omsagent-devs

File: build/Makefile [hunk 1]
```diff
@@ -196,6 +204,12 @@ ifeq ($(ULINUX),1)
 endif
 
 #--------------------------------------------------------------------------------
+# Build AUOMS project
+
+$(AUOMS_TARGET_DIR)/auoms-bundle-test.sh :
```
Comment (NarineM) [inline build/Makefile]: What is the meaning of test in the auoms bundle installer name?

File: installer/bundle/bundle_skel.sh [hunk 2]
```diff
@@ -817,6 +834,21 @@ case "$installMode" in
                     [ $TEMP_STATUS -ne 0 ] && BUNDLE_EXIT_STATUS="$TEMP_STATUS"
                 fi
             done
+            for i in bundles/*-bundle-test.sh; do
+                # If filespec didn't expand, break out of loop
+                [ ! -f $i ] && break
+
+                # It's possible we have a test file without bundle; if so, ignore it
+                BUNDLE=`basename $i -bundle-test.sh`
```
Comment (NarineM) [inline installer/bundle/bundle_skel.sh]: What is the  test t file used for? And where the test bundle sh is being built?

File: installer/bundle/bundle_skel.sh [hunk 3]
```diff
@@ -817,6 +834,21 @@ case "$installMode" in
...
+                [ ! -f bundles/${BUNDLE}-*universal.*.sh ] && continue
+
+                ./$i
+                if [ $? -eq 0 ]; then
+                    ./bundles/${BUNDLE}-*universal.*.sh --install $FORCE $restartDependencies
```
Comment (NarineM) [inline installer/bundle/bundle_skel.sh]: I do not seem to find $restartDependencies  defined anywhere, did you miss it during merge?

File: installer/bundle/bundle_skel.sh [hunk 4]
```diff
@@ -885,6 +917,21 @@ case "$installMode" in
...
+            # It's possible we have a test file without bundle; if so, ignore it
+            BUNDLE=`basename $i -bundle-test.sh`
+            [ ! -f bundles/${BUNDLE}-*universal.*.sh ] && continue
+
+            ./$i
+            if [ $? -eq 0 ]; then
+                ./bundles/${BUNDLE}-*universal.*.sh --upgrade $FORCE $restartDependencies
+                TEMP_STATUS=$?
+                [ $TEMP_STATUS -ne 0 ] && BUNDLE_EXIT_STATUS="$TEMP_STATUS"
+            fi
```
Comment (NarineM) [inline installer/bundle/bundle_skel.sh]: do you need to echo something when it falls in else block?

File: installer/conf/sudoers [hunk 5]
```diff
@@ -31,6 +31,9 @@ omsagent ALL=(ALL) NOPASSWD: /opt/microsoft/omsagent/ruby/bin/ruby /opt/microsof
 #run hdinsightmanifestreader.rb
 omsagent ALL=(ALL) NOPASSWD: /opt/microsoft/omsagent/ruby/bin/ruby /opt/microsoft/omsagent/bin/hdinsightmanifestreader.rb
 
+# Enable nxOMSAuditdResource
+omsagent ALL=(ALL) NOPASSWD: /opt/microsoft/omsconfig/Scripts/OMSAuditdPlugin.sh
```
Comment (NarineM) [inline installer/conf/sudoers]: per oms agent naming convention the bundle names start with lower case (it is actually all lower case , example omsagent-1.3.0-25.universal.deb.x64.sh ).

File: build/configure [hunk 6]
```diff
@@ -5,6 +5,7 @@ base_dir=`(cd ..; pwd -P)`
 scxomi_dir=`(cd ../../omi/Unix; pwd -P)`
 scxpal_dir=`(cd ../../pal; pwd -P)`
 scxcim_dir=`(cd ../../opsmgr; pwd -P)`
+auoms_dir=`(cd ../../auoms; pwd -P)`
```
Comment (NarineM) [inline build/configure]: you mentioned that there are two packages that need to sit under /data folder . I do not see them being installed through configure or it is done somewhere else?

File: source/code/plugins/auditd_plugin_lib.rb [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+            for ridx in 1..record["record-count"]
+                rname = "record-data-"+(ridx-1).to_s
+                if !record.has_key?(rname)
+                    @log.error "Transformation of Auditd Plugin input failed; Missing field '" + rname + "'"
+                    return nil
+                end
+                rdata = Yajl::Parser.parse(record[rname])
+                rdata["Timestamp"] = OMS::Common.format_time(record["AuditID"].to_f)
+                rdata["AuditID"] = record["AuditID"]
```
Comment (NarineM) [inline source/code/plugins/auditd_plugin_lib.rb]: You might want to check for nill value here. I remember there was a limitation on oms service side on processing data with null entities

File: source/code/plugins/auditd_plugin_lib.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
+require 'yajl'
```
Comment (NarineM) [inline source/code/plugins/auditd_plugin_lib.rb]: Please add unittests. I could not find any. you can look into other ruby plugin unittests that use mock objects for testing.

File: build/Makefile (hunk 1 above)
Comment (taglines) [inline build/Makefile]: auoms-bundle-test.sh is not the bundle. It is the script that checks to see if the bundle can be installed. It serves the same purpose as the *-oss-test.sh scripts.

File: installer/bundle/bundle_skel.sh (hunk 3 above)
Comment (taglines) [inline installer/bundle/bundle_skel.sh]: The value of restartDependencies is set at line 482.

File: installer/bundle/bundle_skel.sh (hunk 2 above)
Comment (taglines) [inline installer/bundle/bundle_skel.sh]: The test script is used at line 845.

File: installer/conf/sudoers (hunk 5 above)
Comment (taglines) [inline installer/conf/sudoers]: OMSAuditdPlugin.sh is NOT a bundle! It is the script used by the nxOMSAUditdPlugin DSC resource to complete the enable/disable "SET" operation.

File: build/configure (hunk 6 above)
Comment (taglines) [inline build/configure]: See lines 45-65 of auoms/build/configure.

File: installer/bundle/bundle_skel.sh (hunk 4 above)
Comment (taglines) [inline installer/bundle/bundle_skel.sh]: No. The "BUNDLE_EXIT_STATUS" value is handled later. And, the bundle itself will report failure to the console.

File: source/code/plugins/auditd_plugin_lib.rb [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+            end
+
+            wrapper = {
+                "DataType"=>"LINUX_AUDITD_BLOB",
+                "IPName"=>"Security",
+                "DataItems"=>records
+            }
+
+            @log.info "Audit Event Processed"
+
```
Comment (sugr4) [inline source/code/plugins/auditd_plugin_lib.rb]: Was this for debugging? do you want this message to be logged everytime an audit event is processed?

File: source/code/plugins/auditd_plugin_lib.rb (hunk 8 above)
Comment (sugr4) [inline source/code/plugins/auditd_plugin_lib.rb]: second on unit tests

File: source/code/plugins/filter_auditd_plugin.rb [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    end
+
+    def shutdown
+        super
+        # This method is called when Fluentd is shutting down.
+        # Use it to free up resources, etc.
+    end
+
+    def filter(tag, time, record)       
+        # Create Security Baseline and Security Baseline Summary blobs based on omsbaseline tool scan & assessment results
```
Comment (sugr4) [inline source/code/plugins/filter_auditd_plugin.rb]: This comment does not seem to fit.

File: build/Makefile (hunk 1 above)
Comment (NarineM) [inline build/Makefile]: I see,  having a comment will clear it out since it has the word "bundle" in the name.

File: installer/bundle/bundle_skel.sh (hunk 4 above)
Comment (NarineM) [inline installer/bundle/bundle_skel.sh]: yes, but the bundle is for handling the error for all the components altogether.

File: installer/conf/sudoers (hunk 5 above)
Comment (NarineM) [inline installer/conf/sudoers]: That is fine, I meant the naming convention used in agent sh and bundle names in general.

Comment (NarineM) [inline installer/conf/sudoers]: I see that other scripts for dsc resources also use upper case convention so it is fine if the purpose is to relate the script with the resources names.

**Comment by NarineM** [comment]
Looks good except the missing tests.

File: build/Makefile (hunk 1 above)
Comment (NarineM) [inline build/Makefile]: I see, I interpreted it as a bundle for installing auoms for testing or mocking. I think having comments would really help. are other oss test scripts have the word "bundle" in the name?

File: installer/bundle/bundle_skel.sh (hunk 2 above)
Comment (NarineM) [inline installer/bundle/bundle_skel.sh]: I still do not understand the meaning of this comment" It's possible we have a test file without bundle; if so, ignore it " , can you rephrase it if the test file is about the bundle-test.sh itself and not some other test input data?

Comment (NarineM) [inline installer/bundle/bundle_skel.sh]: ok, I thought test file here means some other input file that is used by the script which is going to look for it :) Nevermind.

File: installer/bundle/bundle_skel.sh (hunk 4 above)
Comment (NarineM) [inline installer/bundle/bundle_skel.sh]: This is fine, seems like other kits like oss handle it same way.

File: source/code/plugins/auditd_plugin_lib.rb (hunk 8 above)
Comment (taglines) [inline source/code/plugins/auditd_plugin_lib.rb]: Added unit test

File: source/code/plugins/filter_auditd_plugin.rb (hunk 10 above)
Comment (taglines) [inline source/code/plugins/filter_auditd_plugin.rb]: The dangers of copy-n-paste. Will remove.
